### ISSUE TITLE
Make it optional to persist VaadinSession in HttpSession

### ISSUE THREAD

**Issue by mstahv** [vaadin/flow#11921]
Vaadin uses standard HttpSession where the UI is stored. For true session replication everything referenced from Vaadin UI objects needs to be serialisable, which makes development trickier. In some cases it would be ok not to serialise the actual Vaadin UI (deep linking would be enough), but saving still some other data, like related to authentication, would be handy. In these case we don't want Vaadin session to be serialised at all (exceptions on log, maybe even preventing serialisation of the wanted details). 

As most Vaadin developers currently don't pay attention to session serialisation, the default should be **not** to persist Vaadin UI to HttpSession at least when serialising the session (breaking change, needs to be documented to release notes). The current behaviour should be configurable with e.g. a servlet parameter.

**Comment by mshabarov** [comment]
@mstahv Vaadin UIs are not serialised by default in the latest versions and can be configured. This is applicable to development mode.
See https://github.com/vaadin/flow/pull/11647/files.

**Comment by mstahv** [comment]
Yup, development mode kind of works in "the correct way" already, so adjusting the behaviour (and adding some configuration properties) should be rather easy now.

**Comment by mstahv** [comment]
@Artur- You wrote the related development mode feature, do you have some ideas about this? Can this replace the current development time feature (if/as we are basically extending the original functionality to production mode)?

In the VaadinSession there are also for example RequestHandlers, which might in some case be unserialisable. The current implementation might be broken in certain cases. Thus we might want to change the functionality a bit.

**Comment by Artur-** [comment]
The current feature should be expanded to "Serialize UIs (yes/no)". 

For development mode the default is "no". 
For production mode the default is "yes". 

Changing the production mode default would break any existing application that uses UI serialization and not help anybody who is not using it. If we want to disable UI serialization by default, a better option would be to set the property in new projects.

**Comment by denis-anisimov** [comment]
@Artur- , the question was different.

At the moment we have an option : **do not serialize UIs** in dev mode.

`VaadinSession` serialization is not the same as UIs serialization since there are other classes which may cause problems with serialization. Currently they are not taken into account. That's why the description is not about `UI` only but about `VaadinSession` as a whole.

A quick look into the `VaadinSession` class revealed two classes which we can't control (other classes are controlled by us and we ensure they are serializable):
- `RequestHandler` : it's possible to add a custom request handled which can be not serializable.
- `Attributes` class: it stores `Object`s as attributes and a developer may store something not serializable as a VaadinSession attribute.

There can be other classes but at least these two are not covered in the existing implementation.

So the question is : _is there any need_ in the existing functionality which disables  _only_ UI serializability in dev mode OR we can replace the existing functionality with the new one implemented as a fix for this ticket: introduce a config parameter (`true` in dev mode and `false` in production mode)  which don't serialize the classes which we don't control. 
These are at least: `UI` , `RequestHandler`, `Attributes` ?

**Comment by Artur-** [comment]
It is super hard to make sure that your UI and all views are serializable.
It is likely quite easy to make sure that any `RequestHandler` and `Attributes` are serializable.
So they are cannot really be compared 1:1.

Anyway, it would be easiest to understand if we drop the whole `VaadinSession`. Then it works like "If you want to retain something, store it in the HTTP session and not the VaadinSession". 

The only problem with this approach was that it technically did not work because of needed fake sessions and locks etc as nobody found a mechanism to actually prevent only selected attributes in the HTTP session from being serialized.

To be clear:
> is there any need in the existing functionality which disables only UI serializability

No. It is implemented that way because it worked and solves the problem (retain login information even though your views are not serializable)

**Comment by denis-anisimov** [comment]
And now I have doubts how it will work for the problem which is solved now since I'm not sure how serialziation is involved there...

Session is somehow deserialized (by Spring  dev tools) and deserialization may fail because of `UI`.  But `UI` is recreated anyway (since this is a reload?).
So everything still works.

Now we don't deserialize `RequestHandler`s and `Attributes`. At all.
So when the session is deserialized it doesn't have _any_ `RequestHandler` and the `Attributes` is empty.

That will break everything as I understand. So it's not the same at all. This ticket doesn't superseeds the existing functionality.
I'm not sure now why do we need this at all: what's the reason to prevent deserialization of `RequestHandler` s and `Attributes` (and overall VaadinSession) if we can't recreate them anyhow and everything will be broken after deserialization ?

This is the first question which I want to have cleared.

And the another question is : should we forget about `RequestHandler` s and `Attributes` and really care _only_ about `UI`s ?

**Comment by mstahv** [comment]
Answer to the "another question": we should investigate how often those are problem. My hunch is that in Artur's tests/prototypes there are no RequestHandlers, but in real world projects we have those. I remember to be using RequestHandlers, in some of my hobby apps, I can try to check if those e.g. consume non-serialisable services.

**Comment by denis-anisimov** [comment]
It has no sense to answer to the second question without answering to the first.

There is dramatic difference between deserialization of `UI` and `VaadinSession`: `UI` is recreated from scratch every time on browser access (loosing all data which is considered  OK) . `VaadinSession` can't be recreated  with losing `RequestHandler`s : everything will be broken.
So we simply can't avoid serialization of `VaadinSession` _IF_ it's intended to be used in Vaadin application.
If the purpose is something else ( I have no idea what, it should have been specified in the ticket description from the very beginning) then it's OK to drop quite many things.

**Comment by denis-anisimov** [comment]
**UPDATE**

The request handlers in the `VaadinSession` are not the same as `RequestHandler`s used in service (as I though initially).
So things won't be broken totally.

As a result on deserialization the session information will be lost and needs to be recreated from scratch.
But then the question is : who will refill the session?
(the methods `addRequestHandler` should be called explicitly any way and most likely they are called at the session initialization even though we can't guarantee that, but the attributes can be set at any moment).

Because without refilling the session the application will work differently before serialization and after.
